Fuzzing: Add test case logging and regression test template

fitzgen opened the pull request:
Example log output when fuzzing now looks like:

```
[2019-11-25T23:48:40Z INFO  wasmtime_fuzzing] Wrote Wasm test case to: /home/fitzgen/wasmtime/crates/fuzzing/target/scratch/7731-0.wasm
[2019-11-25T23:48:40Z DEBUG wasmtime_fuzzing] Disassembling with wasm2wat: "wasm2wat" "--no-check" "--enable-all" "/home/fitzgen/wasmtime/crates/fuzzing/target/scratch/7731-0.wasm" "-o" "/home/fitzgen/wasmtime/crates/fuzzing/target/scratch/7731-0.wasm"
[2019-11-25T23:48:40Z INFO  wasmtime_fuzzing] WAT disassembly:
    (module)
[2019-11-25T23:48:40Z INFO  wasmtime_fuzzing] Also wrote test case's WAT disassembly to: /home/fitzgen/wasmtime/crates/fuzzing/target/scratch/7731-0.wat
[2019-11-25T23:48:40Z INFO  wasmtime_fuzzing] If this fuzz test fails, copy `/home/fitzgen/wasmtime/crates/fuzzing/target/scratch/7731-0.wasm` to `wasmtime/crates/fuzzing/tests/regressions/my-fuzzing-regression.wasm` and add the following test to `wasmtime/crates/fuzzing/tests/regressions.rs`:
    ```
    #[test]
    fn my_fuzzing_regression_test() {
        let wasm = include_bytes!("./regressions/my-fuzzing-regression.wasm");
        oracles::instantiate(wasm, CompilationStrategy::Auto)
    }
    ```
==7731== ERROR: libFuzzer: deadly signal
    #0 0x5610df911c91 in __sanitizer_print_stack_trace /rustc/f1b882b55805c342e46ee4ca3beeef1d1fa2044b/src/llvm-project/compiler-rt/lib/asan/asan_stack.cc:86:3
    #1 0x5610e3bf0219 in fuzzer::PrintStackTrace() /home/fitzgen/.cargo/git/checkouts/libfuzzer-sys-e07fde05820d7bc6/2542ae4/libfuzzer/FuzzerUtil.cpp:205
    #2 0x5610e3bf3c22 in fuzzer::Fuzzer::CrashCallback()
    /home/fitzgen/.cargo/git/checkouts/libfuzzer-sys-e07fde05820d7bc6/2542ae4/libfuzzer/FuzzerLoop.cpp:232
    <snip>
```

alexcrichton commented on crates/fuzzing/src/lib.rs:
```diff
@@ -1,2 +1,176 @@
+///
+/// Also returns the disassembly as a string.
+fn wasm2wat(input: &Path, output: &Path) -> Result<String, Error> {
+    let mut cmd = Command::new("wasm2wat");
```
Could this perhaps use [`wasmprinter`](https://crates.io/crates/wasmprinter) to avoid external reliance on commands?

fitzgen replied:
Good call!

alexcrichton commented on fuzz/fuzz_targets/compile.rs:
```diff
@@ -1,14 +1,20 @@
 #![no_main]
```
This file became accidentally executable I believe

fitzgen replied:
I have no idea why my machine keeps doing this T.T

alexcrichton commented on crates/fuzzing/src/lib.rs:
````diff
@@ -1,2 +1,176 @@
+```
+#[test]
+fn my_fuzzing_regression_test() {{
+    let {wasm_var} = include_bytes!(\"./regressions/my-fuzzing-regression.wasm\");
````
Ideally our regression tests would be `*.wat` files, so would it be possible to generate tests that use the `*.wat` file instead?

fitzgen replied:
Yeah, there are possibly cases where we panic on invalid wasm or something, but for most test cases we should definitely be able to do this.

fitzgen commented:
Ok! I've addressed your feedback. When the test case that causes the failure can successfully be disassembled to WAT, we get logs like this:

```
[2019-11-26T18:48:46Z INFO  wasmtime_fuzzing] Wrote WAT disassembly to: /home/fitzgen/wasmtime/crates/fuzzing/target/scratch/8437-0.wat
[2019-11-26T18:48:46Z INFO  wasmtime_fuzzing] If this fuzz test fails, copy `/home/fitzgen/wasmtime/crates/fuzzing/target/scratch/8437-0.wat` to `wasmtime/crates/fuzzing/tests/regressions/my-regression.wat` and add the following test to `wasmtime/crates/fuzzing/tests/regressions.rs`:

    ```
    #[test]
    fn my_fuzzing_regression_test() {
        let data = wat::parse_str(
            include_str!("./regressions/my-regression.wat")
        ).unwrap();
        oracles::instantiate(data, CompilationStrategy::Auto)
    }
    ```
```

If the test case cannot be disassembled to WAT, then we get logs like this:

```
[2019-11-26T18:48:46Z INFO  wasmtime_fuzzing] Wrote Wasm test case to: /home/fitzgen/wasmtime/crates/fuzzing/target/scratch/8437-0.wasm
[2019-11-26T18:48:46Z INFO  wasmtime_fuzzing] Failed to disassemble Wasm into WAT:
    Bad magic number (at offset 0)

    Stack backtrace:
        Run with RUST_LIB_BACKTRACE=1 env variable to display a backtrace

[2019-11-26T18:48:46Z INFO  wasmtime_fuzzing] If this fuzz test fails, copy `/home/fitzgen/wasmtime/crates/fuzzing/target/scratch/8437-0.wasm` to `wasmtime/crates/fuzzing/tests/regressions/my-regression.wasm` and add the following test to `wasmtime/crates/fuzzing/tests/regressions.rs`:

    ```
    #[test]
    fn my_fuzzing_regression_test() {
        let data = include_bytes!("./regressions/my-regression.wasm");
        oracles::instantiate(data, CompilationStrategy::Auto)
    }
    ```
```

alexcrichton commented:
👍
